### PR TITLE
Constify more API about X509_NAME & X509_NAME_entry

### PR DESCRIPTION
**Pull request by FdaSilvaYY**
Second try ... ;)

### CODE REVIEW

File: crypto/x509/x509name.c [hunk 1]
```diff
@@ -91,6 +91,15 @@ X509_NAME_ENTRY *X509_NAME_get_entry(const X509_NAME *name, int loc)
         return (sk_X509_NAME_ENTRY_value(name->entries, loc));
 }
 
+X509_NAME_ENTRY *X509_NAME_get0m_entry(X509_NAME *name, int loc)
+{
+    if (name == NULL || sk_X509_NAME_ENTRY_num(name->entries) <= loc
+        || loc < 0)
+        return NULL;
+    else
```
Comment (richsalz) [inline crypto/x509/x509name.c]: i hate else after return.

File: crypto/x509/x509name.c [hunk 2]
```diff
@@ -338,20 +347,34 @@ int X509_NAME_ENTRY_set_data(X509_NAME_ENTRY *ne, int type,
     return (1);
 }
 
-ASN1_OBJECT *X509_NAME_ENTRY_get_object(const X509_NAME_ENTRY *ne)
+const ASN1_OBJECT *X509_NAME_ENTRY_get0_object(const X509_NAME_ENTRY *ne)
 {
     if (ne == NULL)
```
Comment (richsalz) [inline crypto/x509/x509name.c]: i don't think we should have these NULL checks here, and below.

File: crypto/x509v3/v3_alt.c [hunk 3]
```diff
@@ -301,7 +301,7 @@ static int copy_email(X509V3_CTX *ctx, GENERAL_NAMES *gens, int move_p)
 {
     X509_NAME *nm;
     ASN1_IA5STRING *email = NULL;
-    X509_NAME_ENTRY *ne;
+    const X509_NAME_ENTRY *ne;
     GENERAL_NAME *gen = NULL;
     int i;
     if (ctx != NULL && ctx->flags == CTX_TEST)
```
Comment (richsalz) [inline crypto/x509v3/v3_alt.c]: as long as you're here, blank line after decls.

File: crypto/x509v3/v3_ncons.c [hunk 4]
```diff
@@ -245,13 +245,13 @@ int NAME_CONSTRAINTS_check_CN(X509 *x, NAME_CONSTRAINTS *nc)
     /* Process any commonName attributes in subject name */
 
     for (i = -1;;) {
-        X509_NAME_ENTRY *ne;
-        ASN1_STRING *hn;
+        const X509_NAME_ENTRY *ne;
+        const ASN1_STRING *hn;
         i = X509_NAME_get_index_by_NID(nm, NID_commonName, i);
```
Comment (richsalz) [inline crypto/x509v3/v3_ncons.c]: blank line after decls

File: crypto/x509v3/v3_utl.c [hunk 5]
```diff
@@ -419,21 +419,21 @@ STACK_OF(OPENSSL_STRING) *X509_REQ_get1_email(X509_REQ *x)
...
     STACK_OF(OPENSSL_STRING) *ret = NULL;
-    X509_NAME_ENTRY *ne;
-    ASN1_IA5STRING *email;
-    GENERAL_NAME *gen;
+    const X509_NAME_ENTRY *ne;
+    const ASN1_IA5STRING *email;
+    const GENERAL_NAME *gen;
     int i;
     /* Now add any email address(es) to STACK */
     i = -1;
```
Comment (richsalz) [inline crypto/x509v3/v3_utl.c]: as long as you're here, merge this with the decl and add a blank line.

**Comment by dot-asm** [comment]
> overall this looks good

I feel that I have to challenge such changes as being *formally* appropriate in the *current* context. Once again, *formally* and in *current* context. Since we don't take binary compatibility lightly, how do we maintain it. Among other things one can formulate pair of trivial rules at source code level:

- there may be *no changes* to public headers between "letter" releases, such as between 1.1.0 and 1.1.0a.
- there may be *only additions* to public headers between minor releases, such as between 1.1.0 and 1.1.1.

As for current context, mind that master is currently 1.1.1, not 1.2.0, which means that proposed changes violate second rule. As for "formal" part. Some rules can be bent, but conditions for bending has to be discussed and agreed upon. Were these actually discussed? Formulated/formalized? For example "constification is deemed acceptable between minor releases as long as constified objects are opaque [to application code]."

**Comment by richsalz** [comment]
I don't disagree with you, @dot-asm.  I was merely commenting on the code itself.  No context, if you will.  Since this changes API's, it's likely to wait until 1.2.0 so this is *on hold* for now.

**Comment by dot-asm** [comment]
In such case you'd have to agree with suggestion to revert 81f9ce1e1965e0e33db6d2391285c4c1b6af0434

**Comment by dot-asm** [comment]
I mean I hereby suggest to revert it, and if so desired reopen corresponding request and put it on hold too.

**Comment by richsalz** [comment]
Will you plus-one the revert?  If so I’ll do that.

**Comment by mattcaswell** [comment]
I'm not sure that changing input parameters to add "const" has any impact on the ABI, or any material impact on the API. Changing output and return parameters does materially impact the API; and since the commit in question does do that: +1 for the revert

**Comment by dot-asm** [comment]
> I'm not sure that changing input parameters to add "const" has any impact on the ABI.

As implied, there are two sides here: *formal* and *practical*. While constification might not change sequence of machine instructions it takes to invoke either original or amended function (that's how you look at it from practical angle, right), *formally* function prototype does change. But let's ask ourselves if the practical angle is actually that practical. You say "I'm not sure", and question is what would it take to turn it to "I ensure that modification is safe [for binary code compiled with original header]"? Since standard doesn't specify such transition what options are there? Examine binary code generated by all involved compilers and their versions (including future ones)? Doesn't sound very practical... Or in other words isn't adhering to *formal* side actually more practical? Paradoxically...

BTW, as for "might not change sequence of machine instructions". Note that const declaration in argument list is a contractual obligation that callee won't modify the referred object. This means that caller would be permitted to refer to object elements prior call and even cache the results for use *after* the call. And this actually affects sequence of machine instructions *around* the invocation. Of course one can argue that binary code compiled with original header may not use this possibility and hence goes safe, but one has to actually say the words and even record them for public to challenge. Binary compatibility is not occasion to rely on "*should*-be-fine" judgement on per-case basis(*). Or at least I don't view it that way. And here is additional angle. Even though it might be possible to work out and formulate some practical rules that could be exercised between minor releases (for example one is suggested above), question still is what would users who care about binary compatibility appreciate more? Running diff on public headers and noting *additions* only? Or observing *modifications* and trying to match them to stated rules?

(*) Note that it doesn't preclude possibility of per-case judgements, just that there ought to be exceptional reasons other than should-be-fine rationale.

**Comment by dot-asm** [comment]
And just in case it wasn't obvious, +1 for revert too.

**Pull request closed without merging**